src/IO/ADIOS/ADIOS2IOHandler.cpp

```diff
@@ -2665,32 +2850,77 @@ namespace detail
             {
                 m_alreadyEnqueued.emplace_back(std::move(task));
             }
-            if (writeAttributes)
+            if (writeLatePuts)
             {
-                for (auto &task : m_attributeWrites)
-                {
-                    m_alreadyEnqueued.emplace_back(
-                        std::unique_ptr<BufferedAction>{
-                            new BufferedAttributeWrite{
-                                std::move(task.second)}});
-                }
-                m_attributeWrites.clear();
+                throw error::Internal(
+                    "ADIOS2 backend: Flush of late writes was requested at the "
+                    "wrong time.");
             }
             m_buffer.clear();
             break;
         }
     }
 
-    void BufferedActions::flush(FlushLevel level, bool writeAttributes)
+    void
+    BufferedActions::flush(ADIOS2FlushParams flushParams, bool writeLatePuts)
     {
+        auto decideFlushAPICall = [this, flushTarget = flushParams.flushTarget](
+                                      adios2::Engine &engine) {
+#if ADIOS2_VERSION_MAJOR * 1000000000 + ADIOS2_VERSION_MINOR * 100000000 +     \
+        ADIOS2_VERSION_PATCH * 1000000 + ADIOS2_VERSION_TWEAK >=               \
+    2701001223
+            bool performDataWrite{};
+            switch (flushTarget)
+            {
+            case FlushTarget::Disk:
+            case FlushTarget::Disk_Override:
+                performDataWrite = true;
+                break;
+            case FlushTarget::Buffer:
+            case FlushTarget::Buffer_Override:
+                performDataWrite = false;
+                break;
+            }
+            performDataWrite = performDataWrite && m_engineType == "bp5";
+
+            if (performDataWrite)
+            {
+                // should we really do this now?
```

Comment:
```suggestion
```

Should we write attributes before PerformDataWrites()?

1. Yes: Then setting another value for the attribute later won't work (see e.g SerialIOTest:)
    ```cpp
        o.setOpenPMDextension(
            1); // this happens intentionally "late" in this test
    ```
2. No: Then attributes would only be written very late, i.e. before `EndStep()`.

Personally, I think if `preferred_flush_target = disk` is selected, then everything that can be written to disk should be, so we should leave it as it is.

Reply:
I think we can revisit this once we change the defaults of frontends and avoid setting defaults of attributes too early.